File: python/coins.py
table = {}

def optimize (N):
	if not N:
		return 0
	else:
		try:
			result = table [N]
			return result
		except (KeyError):
			temp = optimize (N//2) + optimize (N//3) + optimize (N//4)
			if temp > N:
				table[N] = temp
				return temp
			else:
				table[N] = N
				return N

while True:
	try:
		N = int (raw_input ())
		print optimize (N)
	except (ValueError, EOFError):
		break

File: cplusplus/coins.cpp
#include <iostream>
#include <map>


using namespace std;


map <int, long long> table;


long long optimize (long long n)
{
	if (!n)
	{
		return 0;
	}
	else
	{
		long long result = table[n];
		if (result)
		{
			return result;
		}
		else
		{
			long long temp = optimize (n/2) + optimize (n/3) + optimize (n/4);
			if (temp > n)
			{
				table[n] = temp;
				return temp;
			}
			else
			{
				table[n] = n;
				return n;
			}
		}
	}
}



int main()
{
	long long n; // n - coin

	while (cin >> n)
	{
		cout << optimize (n) << endl;
	}
	return 0;
}
